Master

porkybrain opened the pull request:
Patched in support for other formats of bit range expression. Be sure and review it well, my python skills are super elementary at this point.

posborne commented on python/cmsis_svd/parser.py:
```diff
@@ -83,12 +84,25 @@ def _parse_field(self, field_node):
+            bitWidth = 1 + (int(m.group(1)) - int(m.group(2)))     
+        elif msb is not None:
+            bitOffset = lsb;
+            bitWidth = 1 + (msb - lsb);
```
Change looks good in content.  A couple style things to comply with PEP8 (https://www.python.org/dev/peps/pep-0008/):

* Variables should be named snake-case (`bitRange` -> `bit_range`)
* Whitespace around operators (`msb = _get_int(...)`).  Assignment for kwargs (e.g. SVDField constructor) is an exception.
* Python doesn't need the semicolons; they should be removed


You can get a quick check here: http://pep8online.com/.  There are some violations (line length, etc.) in the existing code but it can give you a quick linter.  There is a pep8 command-line checker as well.

kblomqvist commented:
Although I highly prefer snake-case over camel-case, I would here follow element naming according to SVD standard.

posborne commented:
@kblomqvist I'll have to override you here.  The variables in Python code should follow the PEP8 standard and not the SVD/XML standard.

porkybrain commented:
this addition of dim support should give me what I need for now. We can talk about adding more later. Again please have a look if I did it right, this is my first real python experience.

posborne commented:
I'm going to spend a little bit more time looking at this change tomorrow.  Initially, I do not like having the freestanding functions that users have to call that mutate parts of the returned data structure.  I think we can improve on the ergonomics.

Although I do not currently explicitly enforce immutability by restricting access to members like `peripherals.registers`, having external functions that muck around with them is something I would prefer to avoid.  Ideally, the full, correct data structure is either present after parsing or computed based on parsed data by methods bound to the device instance.

porkybrain commented:
From an efficiency standpoint you are completely right. `duplicate_array_of_registers` should meet your criteria of not messing with the underlying tree. We could make it a member function of SVDRegister returning the expanded list of registers or an empty list if the SVDRegister is not an array of registers.

As for the removal of registers and fields named "reserved", if it were my project I would add a parameter to `for_xml_file` called `removeReserved` and default it to true.

porkybrain commented:
I hope this is more in line with your preferred ergonomics. I also added cpu support because I needed it and added support for parsing 'true' and 'false' as ints. I also added a fall back in case a `<dim>` is present but no `<dimIndex>` or use a `n1-n2` range notation rather than a comma seperated list.

posborne commented:
Yeah, this is looking much better.  I think we can improve on it further (with regards to the interface), but I'm fine with merging this as-is.
